src/Tmds.Ssh/PrivateKeyCredential.cs

```diff
@@ -1,20 +1,94 @@
 // This file is part of Tmds.Ssh which is released under MIT.
 // See file LICENSE for full license details.
 
+using System.Security.Cryptography;
+
 namespace Tmds.Ssh;
 
-public sealed class PrivateKeyCredential : Credential
+public class PrivateKeyCredential : Credential
 {
-    internal string FilePath { get; }
+    internal string Identifier { get; }
+
+    private Func<CancellationToken, ValueTask<Key>> LoadKey { get; }
 
-    internal Func<string?> PasswordPrompt { get; }
+    public PrivateKeyCredential(string path, string? password = null, string? identifier = null) :
+        this(path, () => password, identifier)
+    { }
 
-    public PrivateKeyCredential(string path, string? password = null) : this(path, () => password)
+    public PrivateKeyCredential(string path, Func<string?> passwordPrompt, string? identifier = null) :
+        this(LoadKeyFromFile(path ?? throw new ArgumentNullException(nameof(path)), passwordPrompt), identifier ?? path)
     { }
 
-    public PrivateKeyCredential(string path, Func<string?> passwordPrompt)
+    // Allows the user to implement derived classes that represent a private key.
+    protected PrivateKeyCredential(Func<CancellationToken, ValueTask<Key>> loadKey, string identifier)
+    {
+        LoadKey = loadKey;
+        Identifier = identifier;
+    }
+
+    private static Func<CancellationToken, ValueTask<Key>> LoadKeyFromFile(string path, Func<string?> passwordPrompt)
+        => (CancellationToken cancellationToken) =>
+        {
+            if (PrivateKeyParser.TryParsePrivateKeyFile(path, passwordPrompt, out PrivateKey? privateKey, out Exception? error))
+            {
+                return ValueTask.FromResult(new Key(privateKey));
+            }
+
+            if (error is FileNotFoundException or DirectoryNotFoundException)
+            {
+                return ValueTask.FromResult(default(Key));
+            }
+
+            throw error;
+        };
+
+    // This is a type we expose to our derive types to avoid having to expose PrivateKey and a bunch of other internals.
+    protected readonly struct Key
+    {
+        internal PrivateKey? PrivateKey { get; }
+
+        public Key(RSA rsa)
+        {
+            PrivateKey = new RsaPrivateKey(rsa);
+        }
+
+        public Key(ECDsa ecdsa)
+        {
+            ECParameters parameters = ecdsa.ExportParameters(includePrivateParameters: false);
```

Comment:
Is this logic duplicated elsewhere when we do the key loading as a file? Maybe this should be a static method on `ECDsaPrivateKey` to consolidate the logic.

Reply:
I don't think we have something similar to this that we might put in a method.

Reply:
I think you are referring to `TryParseOpenSshEcdsaKey`. I'll take a closer look and see if it makes sense to share some code.

Reply:
I took a look and it's hard to re-use something here because these methods are dealing with different inputs. `TryParseOpenSshEcdsaKey` starts with a key identifier, and a curve name, whereas here we start with an `ECDsa`.